remove use of shared_future, recursive_mutex in HSA runtime

jeffdaily opened the pull request:
Lower the overhead of HSA runtime by removing the use of shared_future.  Its functionality was redundant with the use of shared_ptr<HSAOp> and the HSA signal wait.

jeffdaily committed:
remove use of shared_future in HSA runtime

jeffdaily commented:
jenkins retest this please

scchan commented:
> jenkins retest this please

@david-salinas could you please restart the test once the CI issue has been resolved?

david-salinas commented:
> > jenkins retest this please
> 
> @david-salinas could you please restart the test once the CI issue has been resolved?

rerunning it now.

srinivamd commented on include/hc.hpp:
```diff
@@ -1305,13 +1305,15 @@ class completion_future {
             __amp_future.wait();
         }
+        if (__asyncOp != nullptr) {
+            if (__amp_future.valid()) throw runtime_exception("completion_future expected async op, had amp", 0);
```
Aren't new lines 1308-1309 and 1312-1313 identical in functionality unless the order of the conditions checked matter? If so, some comment would be helpful.
If not, would it be ok to factor it out as: 
`if (__amp_future.valid() &&  (__asyncOp != nullptr)) {`
`    throw runtime_exception("completion_future expected amp, had async op", 0);`
`}`

It could reduce indent of main code.

jeffdaily committed:
remove deprecated HSAQueue::copy_ext due to inf recursion

Using -Wall compilation flag, this function was reported to recurse
infinitely. This function was already marked for removal.

jeffdaily committed:
remove deprecated HSAQueue::copy_ext declaration

jeffdaily committed:
do not use recursive_mutex

jeffdaily commented:
This PR is now updated to further improve runtime overhead.  The use of `recursive_mutex` has been removed.  In many internal APIs where a mutex was assumed to lock, a `NoLock` or `_no_lock` variant has been implemented for use in the methods that already had acquired the lock.

Further, a deprecated API `HSAQueue::copy_ext` was removed.  Compiling with `-Wall` detected infinite recursion in the old API.  Had any client code been calling this function, we would have heard about it by now due to this error.  It was safer to remove it than to continue to leave it deprecated.

These changes result in ~2-7% performance improvement in certain PyTorch benchmarks.

AlexVlx reviewed (approved):
LGTM, thanks.

scchan commented:
LGTM, waiting for the build to finish

jeffdaily commented:
@david-salinas could you please restart the test once the CI issue has been resolved?

jeffdaily commented:
Only one test failure.  Is this a known flaky test or not?
```
Failing Tests (1):
    HCC :: Unit/AmpMath/amp_math_expf_precise_math.cpp

  Expected Passes    : 671
  Expected Failures  : 44
  Unexpected Failures: 1
```

jeffdaily commented:
I ran that particular test on my local system in a loop 1000 times and it failed 13 times.  Perhaps the failure has more to do with the random distribution than any code changes?
```
for i in `seq 1000`; do ./tests/Unit/AmpMath/Output/amp_math_expf_precise_math.cpp.tmp.out; done
sum !< ERROR_THRESHOLD --- 0.104488 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.111008 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.109825 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.107237 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.102685 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.105320 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.102300 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.100736 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.100824 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.101665 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.102851 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.100173 !< 0.100000
sum !< ERROR_THRESHOLD --- 0.102761 !< 0.100000
```

jeffdaily committed:
HSAOp dispose all but the signal after wait

This commit restores the previous behavior of freeing resources
after waiting on a signal. This behavior was changed by
aaa7da165861023e127047795d9794efcfdbff18
in order to maintain thread safety.  Thread safety has since been
restored using std::atomic_flag.

jeffdaily commented:
288fa876a93ddf8e56df4ca9e7bdde5406977e2e 
restores the previous behavior of freeing resources after waiting on a signal. This behavior was changed by
aaa7da1
in order to maintain thread safety.  Thread safety has since been restored using `std::atomic_flag`.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4890,6 +4894,7 @@ inline void
 HSABarrier::wait() {
     // Wait on completion signal until the barrier is finished
     hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+    dispose();
```
Calling `dispose()` in `wait()` doesn't seem right... 
If `wait()` is not called before the `dtor()` of HSAOp, will it cause a memory leak?
Would it be better to call `dispose()` from the `dtor()`? Then, you probably won't need the extra atomic flag since it's supposed to be destroyed by the last owning thread?

jeffdaily replied:
Admittedly, this is the least straightforward part of this PR.  Yes, `wait()` must be called before calling `dispose()`.  This is done in all `dtor()`s.  The reason for calling `dispose()` inside the `wait()` is to decrement the ref counts of stored dep ops as soon as possible.  This is what we were doing before this PR, but this PR accidentally changed that behavior to only decrement dep ops in the `dtor()`.  `dispose()` will free everything but the `_signal` --- the final `_signal` cleanup was moved to the `HSAOp` `dtor`.  `dispose()` should only be called by a single thread, that's why the `atomic_flag` was added.  You're right that the `dtor()` will only be destroyed by the last calling thread, and it will `wait()` just in case.  But any thread can call `wait()` and we should clean up resources there greedily.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -5258,42 +5114,21 @@ void HSAOp::registerSignalCallback() {
-    return _wait_complete_status;
+    if (_signal.handle) {
+        hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+        dispose();
```
I think `dispose()` should be called at the end of this function regardless of whether `_signal.handle` is valid.  Then in `~HSACopy()`,  only `wait()` would be needed and `dispose()` becomes redundant.

jeffdaily replied:
That might be true in this case of `HSACopy`, but isn't so for `HSADispatch` (see comment below).  I agree that in some cases the call to `dispose()` becomes redundant, but the `atomic_flag` at the top of each `dispose()` implementation should be quite cheap.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -763,7 +763,7 @@ class HSAOp : public Kalmar::KalmarAsyncOp {
     activity_prof::ActivityProf _activity_prof;
 
-    hsa_status_t _wait_complete_status;
+    std::atomic_flag _dispose_flag;
```
I think it would be clearer to replace the type of this flag with `std::once_flag`.  Then, we can replace all the calls to `dispose()` with `std::call_once(_dispose_flag, dispose)`.

jeffdaily replied:
The overhead of the `std::atomic_flag` is less than the alternative.  I tried to make it clear with the comment at the top of each `dispose()` implementation.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4660,54 +4604,27 @@ HSADispatch::dispatchKernel(hsa_queue_t* lockedHsaQueue, const void *hostKernarg
-
-        DBOUT (DB_MISC, "complete!\n");
+        hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+        dispose();
```
`dispose()` should be called regardless of whether the signal is valid.  Then, the call to `dispose` in `~HSADispatch()` could be removed.

jeffdaily replied:
If we ever go back to not having every dispatch create a signal, we cannot unconditionally `dispose()` here.  In fact, going back to signal-less dispatches will be an easier change after this PR, so I propose keeping this as-is for now.
